### ISSUE TITLE
aws-cdk-lib: Expose the ability to set a stack timeout to CloudFormation

### ISSUE THREAD

**Issue by damnhandy** [aws/aws-cdk#24792]
### Describe the feature

[Issue 2125](https://github.com/aws/aws-cdk/issues/2125) requested a timeout option to the CDK in order to express the how long a stack should wait before it times out. It was stated that[ CloudFormation doesn't support this](https://github.com/aws/aws-cdk/issues/2125#issuecomment-479372418), but this statement is incorrect. The AWS CLI and SDKs have had the ability to express a stack timeout by setting the [--timeout-in-minutes](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/create-stack.html) on the stack being deployed. Frameworks such as Sceptre are able to leverage this feature and its [stack config takes a timeout argument](https://docs.sceptre-project.org/4.0.2/docs/stack_config.html#stack-timeout). This would be an incredibly useful addition to the CDK. 

A comment in [issue 2125](https://github.com/aws/aws-cdk/issues/2125#issuecomment-552351375) states that this is solved by [#4572](https://github.com/aws/aws-cdk/pull/4572). It is curious why a solution involving custom resources has been deemed a solution to setting a stack timeout when the original issue cited deploying custom resources as the reason for expressing a timeout. In reality, setting a timeout has to happen in advance of deploying a stack. 

### Use Case

There are several use cases where this feature is valuable. Quite often, the default time outs are simply too long.  In cases where an ECS task has a code issue with a container image using a tag such as `latest` (I know, not the best practice, but some insist) and you get into a rollback loop pulling `latest` again. Another use case were things go south is when users mess up Route53 configuration for ACM. In this case, the timeout takes several hours. Being able to have CloudFormation to bail earlier would help a lot.  

### Proposed Solution

The ability to set a stack timeout is not limited to the AWS CLI, but it is also possible to set this via the [AWS SDKs](https://docs.aws.amazon.com/AWSJavaScriptSDK/v3/latest/clients/client-cloudformation/interfaces/createstackcommandinput.html) as well. I would like to propose the ability to express a timeout both on a per-stack basis, as well as at the CDK App level. When a default timeout is specified on a CDK App:

```typescript
const app = new cdk.App({ defaultStackTimeout: Duration.minutes(20)} );
```
Optionally, a default stack time could be expressed by the CLI:

```bash
npx cdk -- deploy --all --stack-timeout-in-minutes 20
```
Each stack managed by the app gets its the same timeout. Both of the above options should yield the same behavior.

An individual stack may also express its own timeout, and override the `defaultStackTimeout` set by the CDK App. 

```typescript
const myStack = new cdk.Stack(app, "MyStack", { stackTimeout: Duration.minutes(10)} );
```
In this case, `myStack` will use a shorter timeout than all the other stacks in the App. 

### Acknowledgements

- [X] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### CDK version used

2.70.0

### Environment details (OS name and version, etc.)

masOS Ventura 13.2.1, RHEL 8

**Comment by shuklsh** [comment]
+ @udaypant @evgenyka 

Another example of a scenario where a custom timeout would have helped troubleshoot faster and save time for developers:

While enabling Sagemaker using CDK, the creation of Sagemaker endpoint would fail without much visibility around what was the root cause (partly also because we use custom images for sagemaker model). While we knew half-way during the stack deployment that the deployment would fail because of bad configuration, there was no way for us to pre-emptively stop the stack from realizing there by resulting in the stack trying 3 times (with exponential backoff) thereby resulting in us having to wait for ~1 hour before we could re-deploy.
This has has a significant impact on us being able to deliver sagemaker usecase for our customer.

@damnhandy fyi

**Comment by peterwoodworth** [comment]
I think this PR should be simple enough to implement if anyone is willing to contribute. We would want to introduce a CLI option for this most likely, and all it would have to do is be passed here to the `createStack` API call https://github.com/aws/aws-cdk/blob/5bdb0128e8181cd3a33a118a1f121f67b0f78b20/packages/aws-cdk/lib/api/deploy-stack.ts#L484-L490

This option would take no effect when direct deployment is not also selected, since this feature is not available on the `createChangeSet` API call

**Comment by mrgrain** [comment]
Yeah we can do this. Like @peterwoodworth pointed out this is only supported by the direct deployment method (using `--method=direct`) in CloudFormation. So we'll have to limited this to the CLI for now and imply `--method=direct` which will cause the loss of progress information.

**Comment by mrgrain** [comment]
Update: The feature is also NOT available in CloudFormation for direct Stack Updates (as opposed to Creation).

**Comment by evgenyka** [comment]
The AWS CLI and SDKs have could express a stack timeout by setting the [--timeout-in-minutes](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/create-stack.html) on Stack Create. If you look into Update Stack (https://docs.aws.amazon.com/cli/latest/reference/cloudformation/update-stack.html), see that the [--timeout-in-minutes] option is not there. 
Since CDK is not distinguishing between creation and update (there is `cdk deploy`), anything that will only work on creation is going to be very awkward for users.

**Comment by GlenTiki** [comment]
@evgenyka Is there any way we as users of CDK can track this required feature in CFN then? Is there even an open feature request against CFN to support a timeout on update? And would it then be possible to keep reopen until downstream has whats needed and its ported up to the CLI? 

Waiting hours for a timeout when you know your update will have failed in 10mins or less is a horrible UX.

**Comment by elijahkim** [comment]
I also had a failed creation event and I've been waiting for the last 3 hours for the rollback to finish so I can delete the stack and start over. It's very unfortunate that a failed configuration can result in such a penalty. 

Just to rant a little more, it's unbelievable that this is still an issue since https://github.com/aws/aws-cdk/issues/2125!

**Comment by peterwoodworth** [comment]
Reopening so we can track fixing once CloudFormation support drops

**Comment by mrgrain** [comment]
### Summary of Current State

The requested [--timeout-in-minutes](https://docs.aws.amazon.com/cli/latest/reference/cloudformation/create-stack.html) option _only_ exists on the `create-stack` API, it does not exists on `update-stack` or `execute-change-set`.

Currently the CDK CLI doesn't differentiate between a stack creation or update - it only has the `cdk deploy` command.
This would lead to a very fractured experience for CDK users, with a lot of explaining and caveating to do.

We are waiting for CloudFormation to support the option on more APIs. Before then the CDK team is not looking to implement this.

**Comment by efekarakus** [comment]
Would it be possible for the CDK CLI to watch for stack updates and if the timeout threshold is reached to invoke [`CancelUpdateStack`](https://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_CancelUpdateStack.html)?

For example, when the timeout is reached:
1. If the stack is deployed, then do nothing.
2. If the stack is being created, then delete it.
3. If the stack is being updated, invoke CancelUpdateStack.

**Comment by JoHuang** [comment]
as @efekarakus mentioned, `3` is the most frequent situation that I've been encountering.
And I need to stay in front of the screen and manually do the timeout job.

**Comment by mrgrain** [comment]
> Would it be possible for the CDK CLI to watch for stack updates and if the timeout threshold is reached to invoke [`CancelUpdateStack`](https://docs.aws.amazon.com/AWSCloudFormation/latest/APIReference/API_CancelUpdateStack.html)?

@efekarakus Absolutely possible. However, as previously stated we think this has significant drawbacks and did not make this a priority at the time:

> We don't want to manage those timeouts on the CDK side, since client-side timeouts are unreliable, so need CloudFormation implementation for update stack.

**Comment by fbritoferreirapf** [comment]
Is this not been updated? I keep getting my jobs stuck as there is no timeout.
